core/src/main/java/org/elasticsearch/common/ParseFieldMatcher.java

```diff
@@ -56,4 +61,35 @@ public boolean isStrict() {
     public boolean match(String fieldName, ParseField parseField) {
         return parseField.match(fieldName, strict);
     }
+
+    /**
+     * Select a named thing from a registry of named things while checking that the name matches the ParseField in the context of this
+     * matcher. Used for looking up parsers for queries, score functions, and aggregations.
+     *
+     * @param registryOf The kind of things stored in the registry. Used to make error messages. Should be singular like "query" instead of
+     *        "queries".
+     * @param registry The actual registry.
+     * @param name The name of the thing to look up.
+     * @param xContentLocation The location in the XContent from which the name was looked up. This is used for error message.
+     * @param nullOnNotFound If true then return null rather than throw an exception when the entry isn't found.
+     * @return The thing being looked up. Never null if nullOk is false.
+     * @throws ParsingException if the named thing isn't in the registry or the name was deprecated and deprecated names aren't supported.
+     */
+    public <T> T selectFromRegistry(String registryOf, Map<String, Tuple<ParseField, T>> registry, String name,
```

Comment:
This doesn't feel like something `ParseFieldMatcher` should do to me. Is there nowhere else this can be?

Also, the JavaDocs here are really quite hard to understand, "thing" is mentioned 6 times. I know this is a fairly abstract concept but we should try to make the JavaDoc a bit easier to consume